src/Interpreters/DDLWorker.cpp

```diff
@@ -964,6 +964,95 @@ bool DDLWorker::tryExecuteQueryOnLeaderReplica(
 }
 
 
+bool DDLWorker::isTaskFinished(const ZooKeeperPtr & zookeeper, const String & node_name)
+{
+    String node_path = fs::path(queue_dir) / node_name;
+
+    String node_data;
+    if (!zookeeper->tryGet(node_path, node_data))
+    {
+        return true;
+    }
+
+    auto log_entry = std::make_unique<DDLLogEntry>();
+    log_entry->parse(node_data);
+
+    ASTPtr ast;
+    try
+    {
+        ParserQuery parser(log_entry->query.data() + log_entry->query.size());
+        ast = parseQuery(parser, log_entry->query, "", 0, context.getSettingsRef().max_parser_depth);
+    }
+    catch (...)
+    {
+        return true;
+    }
+
+    auto * ast_on_cluster = dynamic_cast<ASTQueryWithOnCluster *>(ast.get());
+    if (!ast_on_cluster)
+    {
+        return true;
+    }
+
+    auto cluster = context.getCluster(ast_on_cluster->cluster);
+    if (!cluster)
+    {
+        return true;
```

Comment:
Hosts of multiple clusters can use single DDDL queue. It protects from race condition only if all hosts have definition of all clusters in config. If multiple clusters don't know about each other, then some host will return true from here and remove queue entry.